File: Day-9/labs/python-security-lab/lab01_scram_producer.py
"""
Lab 01-produce with SASL/SCRAM (PLAIN or SCRAM listener).

Requires client-scram.properties and SCRAM user on broker.

Run:
    python lab01_scram_producer.py
    python lab01_scram_producer.py localhost:9093 orders
"""

from __future__ import annotations

import sys

from kafka import KafkaProducer
from kafka.errors import KafkaError

from config import DEFAULT_BOOTSTRAP


def main() -> None:
    bootstrap = sys.argv[1] if len(sys.argv) > 1 else DEFAULT_BOOTSTRAP
    topic = sys.argv[2] if len(sys.argv) > 2 else "orders"
    username = sys.argv[3] if len(sys.argv) > 3 else "alice"
    password = sys.argv[4] if len(sys.argv) > 4 else "secret"

    producer = KafkaProducer(
        bootstrap_servers=bootstrap,
        security_protocol="SASL_PLAINTEXT",
        sasl_mechanism="SCRAM-SHA-512",
        sasl_plain_username=username,
        sasl_plain_password=password,
        value_serializer=lambda v: v.encode("utf-8"),
    )

    try:
        future = producer.send(topic, value="hello-from-python-scram")
        meta = future.get(timeout=10)
        print(f"Sent to {meta.topic} partition={meta.partition} offset={meta.offset}")
    except KafkaError as exc:
        print(f"Failed: {exc}")
        sys.exit(1)
    finally:
        producer.close()


if __name__ == "__main__":
    main()

File: Day-9/labs/java-security-lab/src/main/java/day9/labs/Lab01ScramProducer.java
package day9.labs;

import java.util.concurrent.TimeUnit;
import org.apache.kafka.clients.producer.KafkaProducer;
import org.apache.kafka.clients.producer.ProducerRecord;
import org.apache.kafka.clients.producer.RecordMetadata;

/**
 * Lab 01 — produce one message with SASL/SCRAM.
 *
 * <p>Run: mvn -q exec:java -Dexec.mainClass=day9.labs.Lab01ScramProducer
 * -Dexec.args="localhost:9093 orders alice secret"
 */
public final class Lab01ScramProducer {

  public static void main(String[] args) {
    String bootstrap = args.length > 0 ? args[0] : KafkaClientConfig.defaultBootstrap();
    String topic = args.length > 1 ? args[1] : "orders";
    String username = args.length > 2 ? args[2] : "alice";
    String password = args.length > 3 ? args[3] : "secret";

    try (KafkaProducer<String, String> producer =
        new KafkaProducer<>(
            KafkaClientConfig.scramProducerProperties(bootstrap, username, password))) {

      ProducerRecord<String, String> record =
          new ProducerRecord<>(topic, "hello-from-java-scram");

      RecordMetadata meta = producer.send(record).get(10, TimeUnit.SECONDS);
      System.out.printf(
          "Sent to %s partition=%d offset=%d%n",
          meta.topic(), meta.partition(), meta.offset());
    } catch (Exception e) {
      System.err.println("Failed: " + e.getMessage());
      e.printStackTrace(System.err);
      System.exit(1);
    }
  }
}
